### PR TITLE
[#5] Task02 - Creating sentence from array

### LINKED ISSUE #5: Task02 - Creating sentence from array
For a given array that contains some strings output the content of those strings separated with whitespace between them, having capitalized first word and ending with ".", e.g. [[" thIs ", "Is"], nil, " My  ", ["   first"], "", :program] should output "This is my first program."

### PR DESCRIPTION
**Pull request by MaraEfimova**
An array with different variables is transformed into a valid string that can only contain numbers, letters, commas (not at the beginning and not at the end) and dash (not at the end). Added error display for specific cases.

**Resolves #5**

### CODE REVIEW

**Review by kaloyan-yordanov — CHANGES_REQUESTED** [review]
@MaraEfimova Few issues here:
- `["  ,This"]` is transformed to`, This.`, and an error should be thrown instead (because of the `,`)
- `["This -  "]` is transformed to`This -.` and an error should be thrown instead (because of the `-`)
- `[]` should have a separate error for empty input, because it is in fact not containing illegal characters
- if you pass `{}` the error should say that the input type is invalid and you expect an Array
- remove `p format_array`
- I don't believe `_` are allowed
- `["asd a", "qwe"]` is returning `Asd aqwe.` which is incorrect, should be `Asd a qwe.`
- is `!format_array.scan(/[a-zA-Z]/).empty?` really needed?
- remove all the `elsif`s, simply rise the errors, if no error is raised - the script will ocntinue to the `format_array = format_array.squeeze(" ")` line anyway

File: task02.rb [hunk 1]
```diff
@@ -0,0 +1,28 @@
+#2. For a given array that contains some strings output the content of those strings separated with whitespace between them, having capitalized first word and ending with ".", e.g. [[" thIs ", "Is"], nil, " My  ", ["   first"], "", :program] should output "This is my first program."
+
+def task02(arr)
+  if arr.class != Array
+    raise "The input type is invalid, expect an array"
+  end
+  if arr.empty?
```
Comment (kaloyan-yordanov) [inline task02.rb:7]: @MaraEfimova There is an unexpected error when the input is `[""]`.

File: task02.rb [hunk 2]
```diff
@@ -0,0 +1,28 @@
+#2. For a given array that contains some strings output the content of those strings separated with whitespace between them, having capitalized first word and ending with ".", e.g. [[" thIs ", "Is"], nil, " My  ", ["   first"], "", :program] should output "This is my first program."
+
+def task02(arr)
+  if arr.class != Array
+    raise "The input type is invalid, expect an array"
```
Comment (kaloyan-yordanov) [inline task02.rb]: @MaraEfimova In general I wanted some sort of error, e.g. `raise ArgumentError, "Having error class is better"`. It is still throwing a RuntimeError, but we should always use an explicit error class.

File: task02.rb [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+  end
+  if arr.empty?
+    raise "The input is empty."
+  end
+
+  format_array = arr.flatten.join(" ").strip!
+  if format_array.empty?
+    raise "Not a valid input. The input had only empty data."
+  end
+  if !format_array.scan(/[^\s\w,-]/).empty? || !format_array.scan(/_/).empty?
```
Comment (kaloyan-yordanov) [inline task02.rb]: @MaraEfimova What are those two validation actually doing? And do we need a separate one for `_`?

**Review by kaloyan-yordanov — CHANGES_REQUESTED** [review]

File: task02.rb [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+def task02(arr)
+  if arr.class != Array
+    raise ArgumentError, "The input type is invalid, expect an array"
+  end
+  if arr.empty?
+    raise ArgumentError, "The input is empty."
+  end
+
+  format_string = arr.flatten.join(" ").strip!
+  if format_string.nil?
```
Comment (kaloyan-yordanov) [inline task02.rb]: @MaraEfimova `["", ""]` is returning `.`, which is incorrect and should rise an error instead... The same applies for `["    "]`, `["   ", "    "]`, etc.

File: task02.rb [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+def task02(arr)
+  if arr.class != Array
+    raise ArgumentError, "The input type is invalid, expect an array"
+  end
+  if arr.empty?
+    raise ArgumentError, "The input is empty."
+  end
+
+  format_string = arr.flatten.join(" ").strip!
```
Comment (kaloyan-yordanov) [inline task02.rb]: @MaraEfimova Be careful what the bang method does and returns. This leads to unexpected behavior (e.g. `['me']` throwing an exception).

File: task02.rb [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+  if format_string.nil?
+    raise ArgumentError, "Not a valid input. The input had only empty data."
+  end
+  if !format_string.scan(/([^\s\w,-]|_)/).empty?
```
Comment (kaloyan-yordanov) [inline task02.rb]: @MaraEfimova What is the purpose of this one?

File: task02.rb [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+    raise ArgumentError, "Not a valid input. Letters, numbers, commas and dash are allowed in logical order."
+  end
+  if !format_string.scan(/\A[,]/).empty?
```
Comment (kaloyan-yordanov) [inline task02.rb]: @MaraEfimova Can't we just get the first symbol instead of using RegEx here?

File: task02.rb [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+    raise ArgumentError, "Not a valid input. A sentence cannot start with a comma. Only letters, numbers and dash are allowed."
+  end
+  if !format_string.scan(/[,-]\z/).empty?
```
Comment (kaloyan-yordanov) [inline task02.rb]: @MaraEfimova Can't we simply get the last symbol and check it insted of using RegEx?

File: task02.rb [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+    raise ArgumentError, "Not a valid input. A sentence cannot end with a comma or dash. Only letters and numbers are allowed."
+  end
+
+  format_string = format_string.squeeze(" ").downcase.strip
+  format_string.sub!(/[a-zA-Z]/) {|string| string[0].upcase}
```
Comment (kaloyan-yordanov) [inline task02.rb:26]: @MaraEfimova Why you are using `sub!` here like this?

**Review by kaloyan-yordanov — CHANGES_REQUESTED** [review]

File: task02.rb (hunk 9 above)
Comment (MaraEfimova) [reply task02.rb:26]: Cause sentence could start with dash, so I need change first enterance of letter to uppercase.

**Review by kaloyan-yordanov — APPROVED** [review]
@MaraEfimova This is good to go, please squash and merge it!

**Pull request merged**